Feature/감정, 태그 감정/태그 통계 조회 구현

Soo186 opened the pull request:
## #️⃣ 연관된 이슈

Closes #24
## #️⃣ 작업 내용

### 주요 변경사항
- 월별 감정/태그 통계 조회 API 구현 (/reports/emotions, /reports/tags)
- 다이어리 태그 생성 및 관리 기능 추가 (최대 3개 제한)
- 태그 유효성 검증 로직 구현
 
### 구현 기능

1.  태그 관리
- 허용된 태그 목록 관리
- 다이어리별 태그 CRUD 기능

2.  감정 통계
- 월별 감정 분포 집계
- 기간별 감정 데이터 없을 시 예외 처리

3. 태그 통계
- 월별 태그 사용 빈도 집계
- 기간별 태그 데이터 없을 시 예외 처리

4. 감정/태그 요약 통계
- 월별 감정/태그 요약 통계 집계


## #️⃣ 테스트 결과

> 코드 변경에 대해 테스트를 수행한 결과를 요약해주세요. 예: 모든 테스트 통과 여부, 새로 작성한 테스트 케이스 등

## #️⃣ 변경 사항 체크리스트

- [ ] 코드에 영향이 있는 모든 부분에 대한 테스트를 작성하고 실행했나요?
- [ ] 필요한 경우, 문서를 작성하거나 수정했나요?
- [ ] 코드 컨벤션에 따라 코드를 작성했나요?
- [ ] 본 PR에서 발생할 수 있는 모든 의존성 문제가 해결되었나요?

## #️⃣ 스크린샷 (선택)
<img width="1127" alt="스크린샷 2025-01-27 오후 7 05 50" src="https://github.com/user-attachments/assets/22ab7aec-df45-451c-bc80-c83133aeca99" />


## #️⃣ 리뷰 요구사항 (선택)

디테일한 부분은 다시 리팩토링 하면서 올릴 예정입니다!

예외처리, 에러코드, 등등 클래스도 새로 리팩토링 하면서, 비즈니스 로직도 공부 다시 하면서 리팩토링 하겠습니다!

## 📎 참고 자료 (선택)

[response_1737971982289.json](https://github.com/user-attachments/files/18556495/response_1737971982289.json)

응답 구조 파일 드립니다!

Soo186 committed:
@Vaild > @Valid 오타 수정

Soo186 committed:
@User -> @Tag로 수정

Soo186 committed:
refactor : @Getter 처리로 코드 간소화

Soo186 committed:
feat: 월별 감정 통계 조회 기능 구현

- 월별 감정 통계 조회 서비스 구현 (EmoStatsService)
- 감정 통계 응답 DTO 추가 (EmotStatsRes)
- 감정별 카운트 집계 및 예외처리 로직 구현

Soo186 committed:
refactor : DB 설정 변경

- DB 이름 통일
- ddl auto : update -> none

Soo186 committed:
feat : Tag 상수 설정

- Tag 설정

Soo186 committed:
feat : 사용 가능 태그 조회 api 추가

-  add TagController

Soo186 committed:
feat: 월별 태그 통계 조회 기능 구현

- 태그 통계 API 엔드포인트 추가 (/reports/tags/{year}/{month})
- 월별 태그 사용 빈도수 집계 기능 구현

Soo186 committed:
feat: 태그 생성 및 관리 기능 구현

- 다이어리 작성 시 태그 생성/저장 기능 추가
- 태그 유효성 검증 로직 구현 (최대 3개, 허용된 태그 목록 검증)

Soo186 committed:
feat: TagErrorCode

Soo186 committed:
refactor : TagErrorCode에 맞게 수정

Soo186 committed:
refactor : Cors 수정

Soo186 committed:
feat: 월별 감정 통계 조회 기능 구현

- 감정 통계 API 엔드포인트 추가 (/reports/emotions/{year}/{month})
- 월별 감정 분포 집계 기능 구현

Soo186 committed:
feat: 월별 감정/태그 통계 요약 조회 기능 구현

- 감정 통계 API 엔드포인트 추가 (/reports/summary/{year}/{month})
- 월별 감정/태그 분포 집계 기능 구현 (요약화면)

Soo186 committed:
Merge branch 'dev' into feature/monthly-emotion-statistics-api

Soo186 committed:
refactor : doc refactor

Soo186 committed:
refactor: 저번달과 이번달의 통계를 한꺼번에 조회

- 월별 통계 조회 API 응답 구조 개선 (MultiMonthSummaryRes)
- 단일 API로 이전/현재 달의 통계 데이터 통합 제공

pangwoo96 commented on src/main/java/com/dailyemotion/statics/service/EmoStatsService.java:
```diff
@@ -0,0 +1,47 @@
+
+    private final DiaryRepository diaryRepository;
+
+    public EmoStatsRes getMonthlyStatistics(int year, int month) {
```
제가 이 메소드에 대해서 궁금한 점이 지금 findByDateBetween의 데이터가 없으면 예외를 던지도록 되어있는데 만약에 사용자가 처음 가입하고 접속을 하면 무조건 데이터가 없는 상태이니까 예외가 반드시 터질 수 밖에 없지 않을까요?

제가 이전에 이런 고민을 한 적이 있는데 저는 데이터가 없을 때 빈 리스트를 반환해서 비워두는 식으로 했었습니다. 혹시 이렇게 만드신 의도나 이유가 있을까요??

Soo186 replied:
제가 코드를 짤 때 데이터가 없는 상황을 '오류'라고 생각해서 예외 처리를 한 것 같습니다. 

하지만 실제로는 데이터가 없는 것이 새로운 사용자에게는 매우 자연스러운 상태이고, 이는 예외 처리가 필요한 상황이 아니었네요. 의무로 예외처리를 하려다가 실수를 범했네요.

앞으로는 예외 처리를 할 때도 "이게 정말 예외 상황인가?"를 고민하는 습관 들이도록 해야겠네요 ㅎ

pangwoo96 commented on src/main/java/com/dailyemotion/statics/service/SummaryStatsService.java:
```diff
@@ -0,0 +1,75 @@
+    // 한 달의 통계를 조회하는 메소드
+    public SummaryStatsRes getMonthlySummary(int year, int month) {
+        // 감정 통계 조회
+        EmoStatsRes emotionStats = emotionStatsService.getMonthlyStatistics(year, month);
```
앞서 질문한 것 처럼 getMonthlyStatistics() 안의 findByDateBetween()가 데이터가 없을 경우 처리가 모호한 것 같습니다.

pangwoo96 commented on src/main/java/com/dailyemotion/statics/service/SummaryStatsService.java:
```diff
@@ -0,0 +1,75 @@
+                .collect(Collectors.toList());
+
+        // 태그 통계 조회
+        TagStatsRes tagStats = tagStatsService.getMonthlyStatistics(year, month);
```
마찬가지로 데이터가 없을 경우 예외가 던져지는데 신규 회원의 경우 무조건 예외가 터지는 것 같습니다

pangwoo96 commented on src/main/java/com/dailyemotion/tag/service/TagService.java:
```diff
@@ -1,31 +1,63 @@
+     */
+    private void validateTags(List<String> tags) {
+        if (tags.size() > 3) {
+            throw new TagException(TagErrorCode.TAG_COUNT_EXCEEDED);
```
TagErrorCode.TAG_COUNT_EXCEEDED 에서 TagErrorCode 부분을 삭제하고 스태틱메소드로 호출하면 가독성이 좀 더 좋을것 같습니다!!

Soo186 replied:
넵 감사합니다! 실컷 만들어놓고 안썼네요 ..ㅎㅎ

Soo186 committed:
refactor: 데이터가 없을 시 예외 처리 대신 빈 데이터 반환하도록 개선

- EmoStatsService와 TagStatsService에서 데이터 없을 때 발생하던 예외 처리 로직 제거
- 신규 사용자의 첫 통계 조회 시에도 자연스러운 사용자 경험 제공

Soo186 committed:
style: TagErrorCode 상수 참조 방식 개선

- TagService 클래스의 에러 코드 참조 방식 ->static import로 적용

pangwoo96 reviewed (approved):
확인했습니다~
